### PR TITLE
feat: Inline editing table widget

### LINKED ISSUE #4183: [Feature] Inline Editing of Table widget
## Summary

As a user, I should be able to configure a table widget to have editable rows. When the configuration is enabled from the property pane, an action will be available against each row, onClick, a form should open with all visible row fields which can be edited. The query/api call to ingest changes from the form should be available in the table widget property pane. This form will always open up as form modal. 

## Motivation

This is an easy implementation/workaround of a data grid that can come in handy to community users

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> Added Support for Inline Table editing

Fixes #4183

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

> Tested locally.
> Added test cases.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: FEATURE/4183-inline-editing-table-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.08 **(-0.05)** | 37.15 **(-0.02)** | 33.92 **(-0.05)** | 55.59 **(-0.05)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :red_circle: | app/client/src/widgets/TableWidget/component/Table.tsx | 61.63 **(0)** | 35 **(-3.89)** | 38.89 **(0)** | 60.98 **(0)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableStyledWrappers.tsx | 84.78 **(-2.12)** | 56.12 **(1.23)** | 75.44 **(-4.19)** | 84.78 **(-2.12)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableUtilities.tsx | 37.95 **(0)** | 23.2 **(-0.69)** | 13.04 **(-3.18)** | 35.8 **(-0.38)**
 :green_circle: | app/client/src/widgets/TableWidget/component/index.tsx | 28.57 **(0)** | 29.11 **(1.84)** | 21.05 **(0)** | 27.03 **(0)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/derived.js | 58.02 **(-0.63)** | 58.93 **(-0.83)** | 45.24 **(0)** | 57.95 **(-0.69)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/index.tsx | 28.25 **(-1.63)** | 9.64 **(-0.16)** | 27.78 **(-2.22)** | 28.22 **(-1.32)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@vibhandikyash great work with the deploy preview here 🥳  💯 

I was wondering if the date-picker column should open up the date-picker popup, the experience will be far better, Alternatively we can also remove that type from inline editing (which I would not prefer)

Also the order of properties should probably be re-vamped: Server sider rendering and total record count should go below inline editing prop. 

The inline editing prop needs to have an helper explaining the feature > "Adds an action button to each row, Allows end user to edit, save or cancel. On save a query/API can be called using the onRowUpdate event"

**Comment by SatishGandham** [comment]
@vibhandikyash The inputs are off by about a pixel, can you check that while I review the PR.
I think its from the -1px margin on the right for input.
Screenshot 2021-09-06 at 10 46 39 AM

**Review by SatishGandham — APPROVED** [review]

**Review by somangshu — APPROVED** [review]
Approving for the tests to run

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by SatishGandham** [comment]
@vivekverma2312, New feature. Needs QA

File: app/client/src/widgets/TableWidget/component/TableUtilities.tsx [hunk 1]
```diff
@@ -711,3 +783,85 @@ export const renderDropdown = (props: {
     </div>
   );
 };
+
+export const renderOperationHeaderCell = () => {
+  return (
+    <TableOperationHeaderCell className="th header-reorder" role="columnheader">
+      <div>Operation</div>
```
Comment (SatishGandham) [inline app/client/src/widgets/TableWidget/component/TableUtilities.tsx:790]: Shouldn't this just be edit instead of operation?

**Review by SatishGandham — DISMISSED** [review]

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this feature and please find the status below,

1.By default the default selected row is set to 0 and when the user bind table widget text widget - {{Table1.seletcedRow}} and edit & saves any of the row then the edited row data is not displayed in the text widget - It displays the default selected row data

https://www.loom.com/share/1cd3229f2fec49348b8fb1620a70e2b7

2.[Enhancement] - Providing an edit option individually to every column to edit details or with some different approach -  https://github.com/appsmithorg/appsmith/issues/7756

**Comment by vibhandikyash** [comment]
@YogeshJayaseelan We have setup the new handle to use the {{Table1.triggeredRow}}. Pl. retest and see this further.

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Verified the fix for {{Table1.triggeredRow}} and it is working as expected

Please find the additional issues raised

3.Change any of the column type to date and passing the computed value as {{currentRow.date}} - Click on edit and click on the date field and choose any of the date from the calendar pop up and click anywhere outside the calendar pop and observe user is unable to save the selected date  

https://www.loom.com/share/916cc8132b71410f938f57a9083b6836

4.Editing any of the row and saving the table widget reloads with blank screen for 1-2 seconds

https://www.loom.com/share/7831feda67544bda825b3f87f519bfca

5.In Short Row Height - The cell editable mode is different from Tall and Default 

Screenshot 2021-09-28 at 4 45 09 PM

**Comment by somangshu** [comment]
@vibhandikyash feedback from the team:
-  The buttons for actions column and edit/cancel can be icon buttons in ghost/outline mode.

**Comment by somangshu** [comment]
@vibhandikyash there is some feedback coming in for this issue, Moving this to needs more info until I have it cleared out

**Comment by riodeuno** [comment]
Other than the fact that we'll need designs for this. We will need a new feature design approach here. The previous details I had added were related to making rows editable (whole table) instead of particular columns. Let me figure out an implementation approach and add it to the document. 
I can foresee a need for some POC. There is also a possibility that we'll be introducing v2 for this, if some other features are incompatible or degrade performance.

**Comment by LeeU1911** [comment]
@vibhandikyash Any reason why you're closing this? How can I use this?

**Pull request closed without merging**